kernel/univ.ml

```diff
@@ -712,328 +704,496 @@ let rec cleanup_universes g =
+            let g = change_node g
+              { root with lt; le; gtge;
+                rank = max root.rank (rank_rest + 1);
+                predicative = List.exists (fun n -> n.predicative) to_merge }
+            in
+
+            (* Inserting shortcuts for old nodes. *)
+            let g = List.fold_left (fun g n ->
+              if Level.equal n.univ root.univ then g else enter_equiv g n.univ root.univ)
+              g to_merge
+            in
+
+            (* Updating g.n_edges *)
+            let oldsz =
+              List.fold_left (fun sz u -> sz+List.length u.lt) 0 to_merge +
+              List.fold_left (fun sz u -> sz+List.length u.le) 0 to_merge
+            in
+            let sz = List.length le + List.length lt in
+            let g = { g with n_edges = g.n_edges + sz - oldsz } in
+
+            (* Not clear in the paper: we have to put the newly
+               created component just between B and F. *)
+            List.rev_append f_reindex (root.univ::b_reindex), g
+        in
+
+        (* STEP 5: reindex traversed nodes. *)
+        List.fold_left use_index g to_reindex
+      end
+    in
+
+    (* STEP 6: insert the new edge in the graph. *)
+    let u = repr g u in
+    let v = repr g v in
+    if u == v then
+      if strict then raise CycleDetected else g
     else
-      let leq = reprleq g arcu in
-	(* is some universe >= u good ? *)
-      let good, bad, b_leq =
-	List.fold_left explore (good, bad, false) leq
+      let g =
+        match strict,
+          CList.mem_f Level.equal v.univ u.lt,
+          CList.mem_f Level.equal v.univ u.le with
+        | _, true, _ | false, _, true -> g
+        | true, false, true ->
+          change_node g
+            { u with lt = v.univ :: u.lt;
+              le = CList.except Level.equal v.univ u.le }
+        | _, false, false ->
+          let u = if strict then { u with lt = v.univ :: u.lt }
+                            else { u with le = v.univ :: u.le }
+          in
+          let g = change_node g u in
+          { g with n_edges = g.n_edges + 1 }
       in
-	if b_leq then
-	  arcu::good, bad, true (* b or true *)
-	else
-	  good, arcu::bad, b    (* b or false *)
-  in
-  let good,_,_ = explore ([arcv],[],false) arcu in
-    good
+      let v, g =
+        if not u.predicative || v.predicative then v, g
+        else
+          let v = { v with predicative = true } in
+          v, change_node g v
+      in
+      if u.klvl <> v.klvl || CList.mem_f Level.equal u.univ v.gtge then g
+      else
+        let v = { v with gtge = u.univ :: v.gtge } in
+        change_node g v
+  with
+  | CycleDetected as e -> raise e
+  | e ->
+    (** Unlikely event: fatal error or signal *)
+    let () = cleanup_universes g in
+    raise e
 
-(* We assume  compare(u,v) = LE with v canonical (see compare below).
-   In this case List.hd(between g u v) = repr u
-   Otherwise, between g u v = []
- *)
+(** Uncomment to debug the cycle detection algorithm. *)
```

Comment:
a trivial point, but for non-emacs users it would be easier if the zone to uncomment had a single `(*` at the beginning and `*)` at the end, not one comment per line.